### DISCUSSION TITLE
How to ensure hexagons cover some polygons?

### DISCUSSION THREAD

**Discussion by rmx90210** [bytesandbrains/h3-pg#152] [Q&A]
Hello friends!

I'm trying to use h3-pg to generate resolution-8 hexagons that **fully cover** U.S. states and territories. This is what my first attempt looked like using `h3_polygon_to_cells()`:

![Screenshot 2024-03-28 at 11 07 33 AM](https://github.com/zachasme/h3-pg/assets/1463553/89ab58e2-684e-40d4-b0e1-d6268ea348d0)

Rather than covering the polygons, this seems to be intersecting hexagon centroids with the target geometries. (I think I just had the wrong understanding of how the function works.) 

I'm wondering if this community can suggest a way to find the set of hexagons that...

1. covers a set of polygons at a given resolution, and
2. minimizes the number of non-intersecting hexagons that need to be dropped after the fact (e.g. if you used `ST_Extent`)

I ask this question so specifically because I tried using `ST_Extent` but it generates millions of non-intersecting artifacts, and the intersect + drop operation to remove them has been running for over 12 hours on my machine 😂 (even with a spatial index).

It is potentially better to buffer the polygons rather than taking their extent? If so, how can I determine what buffer distance is guaranteed to cover the polygons for a given H3 resolution?

Thank you all in advance for any advice on this!

**Comment by rmx90210** [comment]
I ended up buffering the target features by 2 * the average edge length of the hexagons given the resolution I was working at. Doubling may have been a bit excessive, but I wanted to err on the side of having too many covering hexagons versus not enough. Then I ran a spatial intersection and dropped any hexagons that didn't actually intersect the target features.

Reference for average edge lengths [here](https://h3geo.org/docs/core-library/restable/#edge-lengths).

**Comment by zachasme** [comment]
Uber actually merged "support for full containment and overlapping modes in polygonToCellsExperimental" (PR: https://github.com/uber/h3/pull/796), back in November, but I'm still waiting for a release so I can include it in my bindings. I believe that will help with this specific use-case.
